src/platform/wpan/nm-wpan-utils.c

```diff
@@ -0,0 +1,286 @@
+		.ack_arg    = &done,
+		.valid_cb   = valid_handler,
+		.valid_arg  = valid_data,
+	};
+
+	g_return_val_if_fail (msg != NULL, -ENOMEM);
+
+	err = nl_send_auto (nl_sock, msg);
+	if (err < 0)
+		return err;
+
+	/* Loop until one of our NL callbacks says we're done; on success
+	 * done will be 1, on error it will be < 0.
+	 */
+	while (!done) {
+		err = nl_recvmsgs (nl_sock, &cb);
+		if (err < 0 && err != -EAGAIN) {
+			_LOGW (LOGD_PLATFORM, "nl_recvmsgs() error: (%d) %s",
+			       err, nl_geterror (err));
+			break;
+		}
+	}
+
+	if (err >= 0 && done < 0)
+		err = done;
+	return err;
+}
+
+static int
+nl802154_send_and_recv (NMWpanUtils *self,
+                        struct nl_msg *msg,
+                        int (*valid_handler) (struct nl_msg *, void *),
+                        void *valid_data)
+{
+	return _nl802154_send_and_recv (self->nl_sock, msg,
+	                                valid_handler, valid_data);
+}
+
+struct nl802154_interface {
+	guint16 pan_id;
+	guint16 short_addr;
+
+	gboolean valid;
+};
+
+static int
+nl802154_get_interface_handler (struct nl_msg *msg, void *arg)
+{
+	struct nl802154_interface *info = arg;
+	struct genlmsghdr *gnlh = nlmsg_data (nlmsg_hdr (msg));
+	struct nlattr *tb[NL802154_ATTR_MAX + 1] = { 0, };
+	static const struct nla_policy nl802154_policy[NL802154_ATTR_MAX + 1] = {
+		[NL802154_ATTR_PAN_ID] =            { .type = NLA_U16 },
+		[NL802154_ATTR_SHORT_ADDR] =        { .type = NLA_U16 },
+	};
+
+	if (nla_parse (tb, NL802154_ATTR_MAX, genlmsg_attrdata (gnlh, 0),
+	               genlmsg_attrlen (gnlh, 0), nl802154_policy) < 0)
+	return NL_SKIP;
+
+	if (tb[NL802154_ATTR_PAN_ID])
+		info->pan_id = le16toh (nla_get_u16 (tb[NL802154_ATTR_PAN_ID]));
+
+	if (tb[NL802154_ATTR_SHORT_ADDR])
+		info->short_addr = le16toh (nla_get_u16 (tb[NL802154_ATTR_SHORT_ADDR]));
+
+	info->valid = TRUE;
+
+	return NL_SKIP;
+}
+
+static void
+nl802154_get_interface (NMWpanUtils *self,
+                        struct nl802154_interface *interface)
+{
+	nm_auto_nlmsg struct nl_msg *msg = NULL;
+
+	memset (interface, 0, sizeof (*interface));
+
+	msg = nl802154_alloc_msg (self, NL802154_CMD_GET_INTERFACE, 0);
+
+	nl802154_send_and_recv (self, msg, nl802154_get_interface_handler, interface);
+}
+
+/*****************************************************************************/
+
+guint16
+nm_wpan_utils_get_pan_id (NMWpanUtils *self)
+{
+	struct nl802154_interface interface;
+
+	nl802154_get_interface (self, &interface);
+
+	return interface.pan_id;
+}
+
+gboolean
+nm_wpan_utils_set_pan_id (NMWpanUtils *self, guint16 pan_id)
+{
+	nm_auto_nlmsg struct nl_msg *msg = NULL;
+	int err;
+
+	g_return_val_if_fail (self != NULL, FALSE);
+
+	msg = nl802154_alloc_msg (self, NL802154_CMD_SET_PAN_ID, 0);
+	NLA_PUT_U16 (msg, NL802154_ATTR_PAN_ID, htole16 (pan_id));
+	err = nl802154_send_and_recv (self, msg, NULL, NULL);
+	return err >= 0;
+
+nla_put_failure:
+	return FALSE;
+}
+
+guint16
+nm_wpan_utils_get_short_addr (NMWpanUtils *self)
```

Comment:
Note that most functions named `nm_platform_${TYPE}_get_${SOMETHING}()` only look into the platform cache -

-- yes, `nm_platform_wifi_get_*()` and `nm_platform_mesh_get_*()` are inconsistent.

Wouldn't a naming scheme `nm_platform_${TYPE}_fetch_${SOMETHING}()` be better to indicate functions that get the information fresh from kernel? I think these two kinds are fundamentally different and the name should reflect that. A get-function that only looks into the cache, will always return the same results (unless the cache changes). But multiple calls to fetch can yield different results and imply a larger overhead.

Compare that also to `nm_platform_link_infiniband_get_properties()`, `nm_platform_link_veth_get_properties()`, and `nm_platform_link_tun_get_properties()`. All these functions first try to look into the platform cache, and if kernel has no support, it uses alternative kernel API to fetch the properties (e.g. sysctl). For that reason,  `nm_platform_link_6lowpan_get_properties()` seems misnamed to me. Well, sure, it also has a fallback, but it still relies entirely on the platform cache to perform the lookup. Maybe just call it `nm_platform_link_6lowpan_get_parent()` and return the ifindex (not boolean)?

Reply:
> note how this code is only called at one place:
> ```
>    short_addr = nm_platform_wpan_get_short_addr (platform, pllink->parent);
>    if (short_addr != G_MAXUINT16) {
>        pan_id = nm_platform_wpan_get_pan_id (platform, pllink->parent);
> ```
> so, it always fetches both. If you would add API like
> ```
> gboolean nm_platform_wpan_fetch_details (NMPlatform *, int ifindex *, guint16 *out_id, guint16 *out_short_addr);
> ```
> it would be less code, but also avoid fetching the same information twice.
 
Yes. On the other hand, the Wi-Fi code has sort of the same problem (there it is for the sake of the legacy WEXT implementation). Also, the separation makes sense for the setters, and having getters consistent with setters makes things looks lest confusing overall.
 
If there's a good reason to address this, then the appropriate fix belongs to the kernel -- it should multicast the property changes, we'd just cache them the way we cache most properties and the getters would just fetch things from the cache. (This also applies to Wi-Fi.)
 
> Note that most functions named `nm_platform_${TYPE}_get_${SOMETHING}()` only look into the platform cache -
>
> -- yes, `nm_platform_wifi_get_*()` and `nm_platform_mesh_get_*()` are inconsistent.
>
> Wouldn't a naming scheme `nm_platform_${TYPE}_fetch_${SOMETHING}()` be better to indicate functions that get the information fresh from kernel? I think these two kinds are fundamentally different and the name should reflect that. A get-function that only looks into the cache, will always return the same results (unless the cache changes). But multiple calls to fetch can yield different results and imply a larger overhead.
 
Perhaps. Given this is not specific to this branch, I won't act on it here. Also, the above note about fixing this in kernel still applies (and would allow us to avoid inventing a weird-ish naming scheme by keeping our get() semantics everywhere).

> Compare that also to `nm_platform_link_infiniband_get_properties()`, `nm_platform_link_veth_get_properties()`, and `nm_platform_link_tun_get_properties()`. All these functions first try to look into the platform cache, and if kernel has no support, it uses alternative kernel API to fetch the properties (e.g. sysctl). For that reason,  `nm_platform_link_6lowpan_get_properties()` seems misnamed to me. Well, sure, it also has a fallback, but it still relies entirely on the platform cache to perform the lookup. Maybe just call it `nm_platform_link_6lowpan_get_parent()` and return the ifindex (not boolean)?
 
No. We do look into the platform cache first too. The whole routinge is basically the same as `nm_platform_link_veth_get_properties`.